### PR TITLE
Add auth to control server requests

### PR DESCRIPTION
**Pull request by RebeccaMahany**
Description:
* Implement control server auth as defined in this PR: https://github.com/kolide/k2/pull/7083
* Switch `valid_until` for notifications to a timestamp
* Set control server hostname + associated settings based on k2 hostname

### CODE REVIEW

File: ee/control/client_http.go
Comment (directionless) [inline ee/control/client_http.go:85]: This would be better as an early return. And, honestly, as something throwing an error.

File: ee/control/client_http.go
Comment (directionless) [inline ee/control/client_http.go:100]: Might need a timeout on this. 

@James-Pickett I wonder if we should bake some kind of helper into krypto. Something like `echelper.Sign(signer1, signer2) (sig1, sig2, error)` or `echelper.Sign(signers ...crypto.Signers) ([]sigs, error)`

**Review by directionless — COMMENTED** [review]
Seems okay to me. Seeing how those signatures are made, I think we could use a helper in `krypto` for it, but I don't know that needs to block this PR. (Though you do need a timeout in hardware signing. )

@seejdev you wrote most of the control server, does this look good to you?

File: ee/control/client_http.go
Comment (RebeccaMahany) [reply ee/control/client_http.go:85]: Ah ok, my understanding was that maybe in the future it'd be possible for the request to be signed only by the hardware keys and not the db keys -- if that's the case, I could check both first to return an error if they're both noop? If not I'll just add an early return here.

File: ee/control/client_http.go
Comment (James-Pickett) [reply ee/control/client_http.go:102]: I've exposed a func in krypto echelper that does signing with time out, accepting duration and interval as parameters. I played around with what @directionless suggested, but it felt weird because ...

what if it returns an error?
which one failed? 
I accept it if only one fails, but if I get back an array of signatures, how do I know which sig matches which key?

so I thought about having custom error messages that says which index of the given sigs failed ... but I feel like that added complexity is > the complexity of just calling sign twice.

I'm open to discussion =)

File: ee/control/client_http.go
Comment (seejdev) [inline ee/control/client_http.go:145]:
```suggestion
	dataReq.Header.Set("Authorization", fmt.Sprintf("Bearer %s", c.token))
```

File: ee/control/client_http.go
Comment (seejdev) [inline ee/control/client_http.go:134]: It may be better to just error out here instead of trying to handle an empty token. The caller needs to get hashes from the data returned by `GetConfig`, so if this function is being called without `GetConfig` called first, the hash is potentially stale or otherwise invalid.

**Comment by seejdev** [comment]
So exciting!

**Comment by seejdev** [comment]
> Seems okay to me. Seeing how those signatures are made, I think we could use a helper in `krypto` for it, but I don't know that needs to block this PR. (Though you do need a timeout in hardware signing. )
> 
> @seejdev you wrote most of the control server, does this look good to you?

LGTM - I have one concern on the `GetSubsystemData` implementation.

I'm glad this fit in relatively cleanly, and great to see @James-Pickett 's crypto work come to fruition here! Nice job!

File: ee/control/client_http.go
Comment (RebeccaMahany) [reply ee/control/client_http.go:134]: Oh yup that absolutely makes more sense, I'll pull this out!

File: cmd/launcher/options.go
Comment (directionless) [inline cmd/launcher/options.go:185]: We should not allow a command line flag.

I'd probably recommend a switch statement, with the default case setting it to disabled.

Comment (RebeccaMahany) [reply cmd/launcher/options.go:185]: We're not allowing a command line flag for the control server -- https://github.com/kolide/launcher/blob/5f2907715481767a07f43dbbcdff45990eafcc32/cmd/launcher/options.go#L94

File: ee/control/client_http.go
Comment (James-Pickett) [inline ee/control/client_http.go:89]: the krypto PR I have open now exposes funcs like

`PublicEcdsaToB64Der`
`SignWithTimeout`

you don't have to use them, but I'll let you know when that PR merges if interested

File: cmd/launcher/options.go
Comment (directionless) [reply cmd/launcher/options.go:185]: `*flDisableControlTLS` I'm not sure we should expose this as a flag.

File: ee/control/client_http.go
Comment (directionless) [reply ee/control/client_http.go:103]: I _suspect_ the common case is going to be consistent:
1. first sig is local, hard fail for that
2. second sig is hardware. Probably don't fail. maybe expose option.

I don't know what's more idiomatic. Make the callers do the work, or write a helper. I'm happy to roll with whatever.

File: cmd/launcher/options.go
Comment (RebeccaMahany) [reply cmd/launcher/options.go:185]: I'll move it to the deprecated section 👍

File: ee/control/client_http.go
Comment (RebeccaMahany) [reply ee/control/client_http.go:89]: Yes definitely!

File: ee/control/client_http.go
Comment (James-Pickett) [reply ee/control/client_http.go:103]: I think an uncommon success condition like "1st of n" should be the responsibility of the consumer .... If I'm using a library and I see a func `SignWithTimeout` that accepts multiple signers I expect an error if any of them fails. I know we are the only consumers of this library, but it just feels weird to me =). So if you good with whatever, lets let launcher handle it.

File: ee/control/client_http.go
Comment (RebeccaMahany) [reply ee/control/client_http.go]: Updated to add an early return with an error here

File: cmd/launcher/options.go
Comment (directionless) [inline cmd/launcher/options.go:197]: If we don't match any of these (a user running launcher outside kolide) we should disable the control server.

File: cmd/launcher/options.go
Comment (directionless) [inline cmd/launcher/options.go:193]:
```suggestion
		// We don't plumb flRootPEM through to the control server, just disable TLS for now
		insecureTLS = true
```

File: cmd/launcher/options.go
Comment (RebeccaMahany) [reply cmd/launcher/options.go:198]: `Control` is already hardcoded to false below (and the corresponding flag is marked as deprecated above) -- but I don't think that flag is actually in use right now. Is there another way you meant for disabling the control server?

Comment (directionless) [reply cmd/launcher/options.go:198]: Hrm. I think there's some out of date bits here. 

The intent, is to run the control server, with those hostnames. If launcher is not talking to one of these, no control server.

I think opts.Control isn't used any more. So in the default case, launcher would try to hit localhost. Which doesn't seem like what we want.

I think either we plumb opts.Control back. or we remove it, and do something with `controlServerURL` being `""`

Comment (RebeccaMahany) [reply cmd/launcher/options.go:198]: @seejdev do you have a suggestion? Otherwise I'm inclined to make `opts.Control` do something.

I think currently it will try to hit localhost by default, so I can also leave it as-is (no behavior change) for this PR and change it up in a subsequent PR once we know how we'd like this to work.

Comment (directionless) [reply cmd/launcher/options.go:198]: I think my bias would be to remove `opts.Control`, and just check for whether `controlServerURL` is empty. No real need for 2 variables.

Comment (directionless) [reply cmd/launcher/options.go:198]: Could probably kick that to the next PR. Once `krypto` merges, we'll want to update the calling functions

**Comment by mouchar** [comment]
When we can expect this change to reach the stable channel? Now, the launcher is mindlessly trying to access https://localhost:3000 every minute.
It interferes with our apps running on this port.

File: ee/control/client_http.go
Comment (RebeccaMahany) [reply ee/control/client_http.go]: Updated to use these!

File: cmd/launcher/options.go
Comment (RebeccaMahany) [reply cmd/launcher/options.go:198]: Discussed offline + updated accordingly -- now we don't run the control service if `controlServerURL` is empty, and also updated `krypto`

**Review by James-Pickett — APPROVED** [review]

**Comment by RebeccaMahany** [comment]
> When we can expect this change to reach the stable channel? Now, the launcher is mindlessly trying to access https://localhost:3000 every minute. It interferes with our apps running on this port.

Hi @mouchar, thanks for bringing this issue to our attention. We will release a patch 0.13.6 to stable today to address it. 🙂

**Comment by directionless** [comment]
> When we can expect this change to reach the stable channel? Now, the launcher is mindlessly trying to access https://localhost:3000 every minute. It interferes with our apps running on this port.

@mouchar I'm so sorry you ran into an issue. We didn't intend for that to go live. As @RebeccaMahany says, we'll get it backed out.

**Comment by directionless** [comment]
@mouchar We have released 0.13.6, which should fix your issue.

File: ee/control/client_http.go
Comment (directionless) [inline ee/control/client_http.go:103]: `hardwareKeys.Public() != nil` is probably a little cleaner

**Review by directionless — COMMENTED** [review]
TLS stuff seems a little weird, not sure I understand it right though

File: cmd/launcher/options.go
Comment (RebeccaMahany) [reply cmd/launcher/options.go:184]: Those options do [two different things](https://github.com/kolide/launcher/blob/main/ee/control/client_http_option.go) for the control service client, though -- `insecureTLS` sets `InsecureSkipVerify` on the client, and `disableTLS` sets the base URL's scheme to http.

I agree that it would be better to not overload `*flInsecureTLS` -- how about I add a separate `InsecureControlTLS` option (not controlled by a command line flag) for this?

Comment (RebeccaMahany) [reply cmd/launcher/options.go]: I went ahead and did it ⬆️ this way in the last commit, let me know what you think or if you'd like changes

**Review by directionless — APPROVED** [review]
Let's try!

File: cmd/launcher/options.go
Comment (directionless) [reply cmd/launcher/options.go]: Ah yes, good catch

**Pull request merged**
